### PR TITLE
Clean up TestDef API, clarify naming

### PR DESCRIPTION
**Pull request by katcharov**
In #1557, I neglected to push (optional) PR fixes before merging (naming, etc.). After merging, I saw I could move the assumption out of the API, which cleaned things up a bit.

### CODE REVIEW

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTestModifications.java [hunk 1]
```diff
@@ -372,32 +366,24 @@ private TestApplicator(
...
-            this.shouldSkip = false;
-            this.reasonToApply = null;
             this.modifiersToApply = modifiersToApply;
         }
 
         private TestApplicator(
                 final TestDef testDef,
                 @NonNull
-                final String reason) {
+                final String skipReason) {
```
Comment (nhachicha) [inline driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTestModifications.java:375]: I don't think `skipReason` is used?

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTestModifications.java [hunk 2]
```diff
@@ -372,32 +366,24 @@ private TestApplicator(
...
             this.testDef = testDef;
-            this.shouldSkip = true;
-            this.reasonToApply = reason;
-            this.modifiersToApply = new ArrayList<>();
+            this.modifiersToApply = Arrays.asList(Modifier.SKIP);
```
Comment (nhachicha) [inline driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTestModifications.java:377]: Maybe ...
```suggestion
            this.modifiersToApply = Collections.singletonList(Modifier.SKIP);
```

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTestModifications.java [hunk 3]
```diff
@@ -372,32 +366,24 @@ private TestApplicator(
                 final TestDef testDef,
                 final List<Modifier> modifiersToApply) {
             this.testDef = testDef;
-            this.shouldSkip = false;
-            this.reasonToApply = null;
             this.modifiersToApply = modifiersToApply;
         }
 
         private TestApplicator(
```
Comment (nhachicha) [inline driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTestModifications.java:372]: Should we remove this ctor and force the caller to provide the modifiers (if any)? One could accidentally invoke this ctor and have all tests skipped...

**Review by nhachicha — COMMENTED** [review]
LGTM minor things ...

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTestModifications.java (hunk 1 above)
Comment (katcharov) [reply driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTestModifications.java:375]: It is present to force developers to provide the reason (it is not shown in evergreen or I think elsewhere, if used in a failed assumption). I added an `assertNotNull`.

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTestModifications.java (hunk 3 above)
Comment (katcharov) [reply driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTestModifications.java]: Good point, I cleaned this up to have one constructor, and this also makes it clearer what is happening in the methods that call the constructor.

**Review by nhachicha — APPROVED** [review]
🎉

**Pull request merged**